components/core/src/clp/ffi/KeyValuePairLogEvent.cpp

```diff
@@ -0,0 +1,174 @@
+            return value.is<FourByteEncodedTextAst>() || value.is<EightByteEncodedTextAst>();
+        case SchemaTreeNode::Type::Str:
+            return value.is<string>() || value.is<FourByteEncodedTextAst>()
+                   || value.is<EightByteEncodedTextAst>();
+        default:
+            return false;
+    }
+}
+
+auto validate_node_id_value_pairs(
+        SchemaTree const& schema_tree,
+        KeyValuePairLogEvent::NodeIdValuePairs const& node_id_value_pairs
+) -> std::optional<std::errc> {
+    std::optional<std::errc> ret_val;
+    std::unordered_map<SchemaTreeNode::id_t, std::unordered_set<std::string_view>>
+            parent_node_id_to_key_names;
+    try {
+        for (auto const& [node_id, value] : node_id_value_pairs) {
+            if (SchemaTree::cRootId == node_id) {
+                ret_val.emplace(std::errc::operation_not_permitted);
+                break;
+            }
+
+            auto const& node{schema_tree.get_node(node_id)};
+            auto const node_type{node.get_type()};
+            if (false == value.has_value()) {
+                // Value is an empty object (`{}`, which is not the same as `null`)
+                if (SchemaTreeNode::Type::Obj != node_type) {
+                    ret_val.emplace(std::errc::protocol_error);
+                    break;
+                }
+            } else if (false == node_type_matches_value_type(node_type, value.value())) {
+                ret_val.emplace(std::errc::protocol_error);
+                break;
+            }
+
+            if (SchemaTreeNode::Type::Obj == node_type
+                && false == is_leaf_node(schema_tree, node_id, node_id_value_pairs))
+            {
+                // Implicit key conflict: A `null` or empty value is given but its descendants
+                // appear in the node ID value pairs
+                ret_val.emplace(std::errc::protocol_not_supported);
+                break;
+            }
+
+            auto const parent_node_id{node.get_parent_id()};
+            auto const key_name{node.get_key_name()};
+            if (parent_node_id_to_key_names.contains(parent_node_id)) {
+                if (parent_node_id_to_key_names.at(parent_node_id).contains(key_name)) {
+                    // Explicit key conflict: the key is duplicated under the same parent
+                    ret_val.emplace(std::errc::protocol_not_supported);
+                    break;
+                }
+            } else {
+                parent_node_id_to_key_names.emplace(parent_node_id, std::unordered_set{key_name});
+            }
+        }
+    } catch (SchemaTree::OperationFailed const& ex) {
+        ret_val.emplace(std::errc::operation_not_permitted);
+    }
+    return ret_val;
+}
+
+auto is_leaf_node(
+        SchemaTree const& schema_tree,
+        SchemaTreeNode::id_t node_id,
+        KeyValuePairLogEvent::NodeIdValuePairs const& node_id_value_pairs
+) -> bool {
+    std::vector<SchemaTreeNode::id_t> dfs_stack;
```

Comment:
Have you considered using a `deque`?

Reply:
In the next PR, you will see that I'm actually using `std::stack` (which is implemented using `deque`) for dfs traversal, which is because the stack element is more complicated and using a vector means I need to provide move/copy semantics.
However, in this scenario, the stack is simply an int. I think the guideline of using `std` container is you should always try to use `vector` unless you have a strong motivation to use other ones. In this particular function, using `vector` have better locality, fewer indirections, and less memory allocation. Therefore, I would prefer to stick to `vector` unless you have other concern 😄

Reply:
Sure, was just asking.